### ISSUE TITLE
Power and CPU utilization metrics are instant and that is why not accurate 

### ISSUE THREAD

**Issue by homm** [vladkens/macmon#10]
I have running MX Power gadget, which introduces significant load about 50% of the one core.

Macmon show this as either zero or 100% load for different time periods, which is not true, since it runs every second and every second average load is about 50%.
<img width="1018" alt="Screenshot 2024-11-14 at 14 30 59" src="https://github.com/user-attachments/assets/4fc04d7e-272f-41a0-bb01-31c9b9203b7c">

This is how the same period it actually looks in [pgauge](https://github.com/homm/pgauge) which uses `powermetrics`:
<img width="507" alt="Screenshot 2024-11-14 at 14 31 09" src="https://github.com/user-attachments/assets/33dc05db-adb2-46e7-9ad7-455ebe9273ec">

**Comment by vladkens** [comment]
Hi, @homm. Thank you for issue.

### Initial investigation

1. Install [frida](https://frida.re/): `pip install frida-tools`
2. Disable SIP and update boot args. See [here](https://github.com/frida/frida/issues/2115#issuecomment-1221297876)
3. Trace `powermetrics`:
  a. Run `powermetrics --samplers cpu_power -i 1000`
  b. In second terminal run:

```sh
frida-trace -i "IOReportCreateSamplesDelta" -i "IOReportCreateSamples" powermetrics
```

<details>
<summary>Output</summary>

```
296 ms  IOReportCreateSamples()
   297 ms  IOReportCreateSamplesDelta()
   297 ms  IOReportCreateSamples()
   298 ms  IOReportCreateSamplesDelta()
   298 ms  IOReportCreateSamples()
   299 ms  IOReportCreateSamplesDelta()
   299 ms  IOReportCreateSamples()
   301 ms  IOReportCreateSamplesDelta()
   301 ms  IOReportCreateSamples()
   302 ms  IOReportCreateSamplesDelta()
   302 ms  IOReportCreateSamples()
   302 ms  IOReportCreateSamplesDelta()
   302 ms  IOReportCreateSamples()
   302 ms  IOReportCreateSamplesDelta()
   302 ms  IOReportCreateSamples()
   304 ms  IOReportCreateSamplesDelta()

  1321 ms  IOReportCreateSamples()
  1322 ms  IOReportCreateSamplesDelta()
  1322 ms  IOReportCreateSamples()
  1323 ms  IOReportCreateSamplesDelta()
  1323 ms  IOReportCreateSamples()
  1325 ms  IOReportCreateSamplesDelta()
  1325 ms  IOReportCreateSamples()
  1327 ms  IOReportCreateSamplesDelta()
  1327 ms  IOReportCreateSamples()
  1327 ms  IOReportCreateSamplesDelta()
  1328 ms  IOReportCreateSamples()
  1328 ms  IOReportCreateSamplesDelta()
  1328 ms  IOReportCreateSamples()
  1328 ms  IOReportCreateSamplesDelta()
  1328 ms  IOReportCreateSamples()
  1329 ms  IOReportCreateSamplesDelta()
  
  2349 ms  IOReportCreateSamples()
  2350 ms  IOReportCreateSamplesDelta()
  2350 ms  IOReportCreateSamples()
  2351 ms  IOReportCreateSamplesDelta()
  2351 ms  IOReportCreateSamples()
  2352 ms  IOReportCreateSamplesDelta()
  2352 ms  IOReportCreateSamples()
  2355 ms  IOReportCreateSamplesDelta()
  2355 ms  IOReportCreateSamples()
  2355 ms  IOReportCreateSamplesDelta()
  2356 ms  IOReportCreateSamples()
  2356 ms  IOReportCreateSamplesDelta()
  2356 ms  IOReportCreateSamples()
  2356 ms  IOReportCreateSamplesDelta()
  2356 ms  IOReportCreateSamples()
  2357 ms  IOReportCreateSamplesDelta()
```

</details>

From output we can see that `powermetrics` calls `IOReportCreateSamples` 8 times in row, then wait second for next run.

4. Verify `macmon` trace:
```sh
frida-trace -i "IOReportCreateSamplesDelta" -i "IOReportCreateSamples" macmon
```

<details>
<summary>Output</summary>

```
   615 ms  IOReportCreateSamples()
   705 ms  IOReportCreateSamples()
   710 ms  IOReportCreateSamplesDelta()
  1638 ms  IOReportCreateSamples()
  1725 ms  IOReportCreateSamples()
  1729 ms  IOReportCreateSamplesDelta()
```

</details>

`macmon` calls `IOReportCreateSamples` two times with 80ms sleep, then wait next tick (as expected).

### Initial conclusion

Well, the problem here is that on first measurement was no real cores usage, but in next measurement cores can be on high usage or again without usage in given period of time. `IOReportCreateSamples` just gives system metrics on current call and then program somehow should interpolate this data to whole interval (1 seconds in both examples). Not sure how `powermetrics` did it, because they only do 8 measurements which takes 10ms from 1000, but results looks better then in `macmon`.

### Further investigation

I created small python script to create p-cores load for 1 sec and wait 1 sec:

<details>
<summary>Code</summary>

```py
import multiprocessing
import time


def loop(core: int):
    # align to start of second
    time.sleep(1 - (time.time() * 1e6 % 1e6) / 1e6)

    gap = 1 # wake / idle time
    while True:
        stime = time.time()
        while time.time() - stime < gap:
            pass

        time.sleep(gap - (time.time() * 1e6 % 1e6) / 1e6)


def main():
    threads = []
    for i in range(4): # p-cores count
        t = multiprocessing.Process(target=loop, args=(i,))
        t.start()
        threads.append(t)

    for thread in threads:
        thread.join()


if __name__ == "__main__":
    main()

```

</details>

And updated `macmon` to archive this results

<table>
  <tr>
    <td align="center">Same with powermetrics</td>
    <td align="center">Same with Activity Monitor</td>
  </tr>

  <tr> 
    <td><img height="300" alt="image" src="https://github.com/user-attachments/assets/f98e6e54-3dc2-485c-95f5-4ebce0bbb15c"></td>
    <td><img height="300" alt="image" src="https://github.com/user-attachments/assets/b0ba2eec-33b6-47bc-bc63-389ab7e05156"></td>
  </tr>
  <tr>
    <td>On each cycle: compare last <code>IOReportCreateSamples</code> with current, then do 7 actual samples in row (without timeout) and calculate average. Keep last sample for next iteration. Sleep.</td>
    <td>On each cycle: compare last <code>IOReportCreateSamples</code> with current. For each <code>samples_count</code> do: sleep for <code>sample_interval / samples_count</code>, collect actual sample; After cycle: keep last sample for next call, calculate average</td>
  </tr>
</table>

Second approach on my taste looks more accurate.

For reference: If samples are made with `-i 100`, both programs show the same results - which generally shows that the problem is not with measurement, but with results interpolation.

In general this problem can be fixed with more frequent sampling, but each sample have visible CPU load, which something not expected from performance monitoring tool.

**Comment by vladkens** [comment]
Released in [v0.4.0](https://github.com/vladkens/macmon/releases/tag/v0.4.0). To update, run:

```sh
brew update && brew upgrade macmon
```
